src/main/java/foundation/privacybydesign/sms/smssender/CMGatewaySender.java

```diff
@@ -0,0 +1,97 @@
+        if (conf == null) {
+            throw new IOException("CM gateway configuration not found");
+        }
+
+        // CM expects 00 to be used as international prefix, instead of the + prefix in the E164 phone number format.
+        if (phone.startsWith("+")) {
+            phone = String.format("00%s", phone.substring(1));
+        } else if (!phone.startsWith("00")){
+            throw new IOException("CM expects internationalized phone numbers");
+        }
+
+        String endpoint = conf.getApiEndpoint();
+        if (!endpoint.startsWith("https://")) {
+            throw new IOException("CM gateway API endpoint should use https");
+        }
+        if (endpoint.endsWith("/")) {
+            endpoint = endpoint.substring(0, endpoint.length() - 1);
+        }
+
+        Map<String,String> parameters = new TreeMap<>();
+        parameters.put("producttoken", conf.getProductToken());
+        parameters.put("from", conf.getFrom());
+        parameters.put("to", phone);
+        parameters.put("body", message);
+        parameters.put("reference", conf.getReference());
+
+        URL url = this.constructURL(String.format("%s/gateway.ashx", endpoint), parameters);
+
+        HttpURLConnection connection = (HttpURLConnection) url.openConnection();
+        connection.setConnectTimeout(SMSConfiguration.getInstance().getSMSSenderTimeout());
+        connection.setReadTimeout(SMSConfiguration.getInstance().getSMSSenderTimeout());
+
+        int responseStatus = connection.getResponseCode();
+        if (responseStatus != 200) {
+            throw new IOException(String.format("CM gateway returned status code %d", responseStatus));
+        }
+
+        InputStream responseBody = connection.getInputStream();
+        String text;
+        try (Scanner scanner = new Scanner(responseBody, StandardCharsets.UTF_8.name())) {
+            text = scanner.useDelimiter("\\A").next();
+        }
+
+        // CM returns empty string when the SMS is successfully sent
+        if (text == null) {
+            throw new IOException("CM response could not be parsed");
+        } else if (!text.equals("")) {
+            throw new IOException(String.format("Error response received from CM: %s", text));
+        }
+    }
+
+    private URL constructURL(String endpoint, Map<String, String> parameters) throws MalformedURLException {
+        StringBuilder builder = new StringBuilder(endpoint);
+        builder.append("?");
+        for (Map.Entry<String,String> parameter : parameters.entrySet()) {
+            // We only need alphabetic url parameter keys, so we reject all others.
+            if (!parameter.getKey().matches("^[a-z]+$")) {
+                throw new MalformedURLException("Invalid URL parameters");
+            }
+            try {
+                String value = URLEncoder.encode(parameter.getValue(), StandardCharsets.UTF_8.name());
+                builder.append(String.format("%s=%s&", parameter.getKey(), value));
```

Comment:
Both this method and `RESTSender.getBytes()` encode query parameters into an URL, is it possible here to reuse that code instead?

Reply:
Yeah I considered that too. Then I should build a separate helper class to construct URLParameters (because that is the overlapping part). We can do so, but I thought it was not worth the effort for such a small piece of code. But if you disagree, I'm happy to add that.

Reply:
Actually, in the RESTSender it is a x-www-form-urlencoded format and in the CMGatewaySender it are URL parameters (which both follow the same format), so calling it URLParameters would maybe also be confusing.